### PR TITLE
port parabricks/indexgvcf to nf-test

### PR DESCRIPTION
**Pull request by famosab**
## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

**Comment by mashehu** [comment]
are the tests passing on a test instance/local gpu cluster?

**Comment by famosab** [comment]
yes (I only tested with docker)

File: modules/nf-core/parabricks/indexgvcf/tests/main.nf.test [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    test("human - gvcf.gz") {
+
+        when {
+            process {
+                """
+                input[0] = [
+                    [ id:'test' ], // meta map
+                    file('https://github.com/nf-core/test-datasets/raw/0f9d382fb977a63114e2ae3d59a320dbedba0311/data/genomics/homo_sapiens/illumina/gvcf/test.genome.g.vcf.gz')
+                    //file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gvcf/test.genome.g.vcf.gz', checkIfExists: true)
```
Comment (jfy133) [inline modules/nf-core/parabricks/indexgvcf/tests/main.nf.test:43]: Remove if not necessary anymore

File: modules/nf-core/parabricks/indexgvcf/tests/main.nf.test [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+        options "-stub"
+
+        when {
+            process {
+                """
+                input[0] = [
+                    [ id:'test' ], // meta map
+                    file('https://github.com/nf-core/test-datasets/raw/0f9d382fb977a63114e2ae3d59a320dbedba0311/data/genomics/homo_sapiens/illumina/gvcf/test.genome.g.vcf.gz')
+                    //file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gvcf/test.genome.g.vcf.gz', checkIfExists: true)
```
Comment (jfy133) [inline modules/nf-core/parabricks/indexgvcf/tests/main.nf.test:92]: As above

File: modules/nf-core/parabricks/indexgvcf/tests/main.nf.test.snap [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                    [
+                        {
+                            "id": "test"
+                        },
+                        "versions.yml:md5,a6c2622b9fd9223dbc349c2a27e9a2e4"
+                    ]
+                ],
+                "1": [
+                    "versions.yml:md5,a6c2622b9fd9223dbc349c2a27e9a2e4"
+                ],
```
Comment (jfy133) [inline modules/nf-core/parabricks/indexgvcf/tests/main.nf.test.snap:15]: This looks funky to me, where is the index file(s) itself?

File: modules/nf-core/parabricks/indexgvcf/main.nf [hunk 4]
```diff
@@ -1,16 +1,17 @@
...
 
     input:
-    tuple val(meta), path(gvcf, stageAs:'')
+    tuple val(meta), path(gvcf)
 
     output:
     // This tool outputs g.vcf.idx if input is uncompressed, g.vcf.gz.tbi if input is compressed
-    tuple val(meta), path("*.g.vcf*") , emit: gvcf_index
-    path "versions.yml"               , emit: versions
+    tuple val(meta), path("*") , emit: gvcf_index
```
Comment (jfy133) [inline modules/nf-core/parabricks/indexgvcf/main.nf:13]: Literally anything seems a bit extreme, wouldn't that also pick up the `versions.yml` too?

**Comment by jfy133** [comment]
Note I'm not familiar with the GPU udptes but mostly looksl ike adding tags to go on the right runner so fine with me from that PoV :+1:

File: modules/nf-core/parabricks/indexgvcf/main.nf (hunk 4 above)
Comment (famosab) [reply modules/nf-core/parabricks/indexgvcf/main.nf:13]: I am confused - what do you mean by picking up the versions too? And why remove the versions output???

File: modules/nf-core/parabricks/indexgvcf/tests/main.nf.test.snap (hunk 3 above)
Comment (famosab) [reply modules/nf-core/parabricks/indexgvcf/tests/main.nf.test.snap:15]: should be right below that, no? called gvcf_index

File: modules/nf-core/parabricks/indexgvcf/main.nf (hunk 4 above)
Comment (jfy133) [reply modules/nf-core/parabricks/indexgvcf/main.nf:13]: You are specifying literally any output file generated by the entire module, is there not a specific suffix you can give your pattern:

```suggestion
path("*.{idx,tbi}")
```

With just `*` your `gvcf_index` may also include the `versions.yml` file, I believe?

File: modules/nf-core/parabricks/indexgvcf/tests/main.nf.test.snap (hunk 3 above)
Comment (jfy133) [reply modules/nf-core/parabricks/indexgvcf/tests/main.nf.test.snap:15]: I see a meta field and `versions.yml`, there is no `.idx` nor `.tbi` listed with an md5sum

image

Comment (famosab) [reply modules/nf-core/parabricks/indexgvcf/tests/main.nf.test.snap:15]: You are completely right - I think I found the issue.

**Comment by famosab** [comment]
Note: the `.idx` file did not produce stable snapshots so I am just checking for the filename at the moment but I am happy to extend with other ideas.

File: modules/nf-core/parabricks/indexgvcf/main.nf [hunk 5]
```diff
@@ -1,16 +1,17 @@
 process PARABRICKS_INDEXGVCF {
     tag "$meta.id"
     label 'process_high'
+    label 'process_gpu'
+    stageInMode 'copy'
```
Comment (famosab) [inline modules/nf-core/parabricks/indexgvcf/main.nf:5]:
```suggestion
    stageInMode 'copy' // needed by the module to work properly - might be removed when this is fixed upstream
```

**Pull request merged**